web_poet/rules.py

```diff
@@ -0,0 +1,352 @@
+from __future__ import annotations  # https://www.python.org/dev/peps/pep-0563/
+
+import importlib
+import importlib.util
+import pkgutil
+import warnings
+from collections import deque
+from operator import attrgetter
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+import attrs
+from url_matcher import Patterns
+
+from web_poet._typing import get_item_cls
+from web_poet.pages import ItemPage
+from web_poet.utils import _create_deprecated_class, as_list, str_to_pattern
+
+Strings = Union[str, Iterable[str]]
+
+PageObjectRegistryTV = TypeVar("PageObjectRegistryTV", bound="PageObjectRegistry")
+
+
+@attrs.define(frozen=True)
+class ApplyRule:
+    """A rule that primarily applies Page Object and Item overrides for a given
+    URL pattern.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[ApplyRule]`` when calling the
+    ``web_poet.default_registry``'s :meth:`~.PageObjectRegistry.get_rules`
+    method.
+
+    You can access any of its attributes:
+
+        * ``for_patterns`` - contains the list of URL patterns associated with
+          this rule. You can read the API documentation of the `url-matcher
+          <https://url-matcher.readthedocs.io/>`_ package for more information
+          about the patterns.
+        * ``use`` - The Page Object that will be **used** in cases where the URL
+          pattern represented by the ``for_patterns`` attribute is matched.
+        * ``instead_of`` - *(optional)* The Page Object that will be **replaced**
+          with the Page Object specified via the ``use`` parameter.
```

Comment:
> I think we should clarify later in the docstring how this parameter affects the mapping. I assume that, if you do not specify instead_of, use is used for all matches of for_patterns, but if you specify instead_of, use is used only for matches of for_patterns whose page object would otherwise be instead_of.

The idea is almost there. I added some notes in the docstrings about this to remove the ambiguity and clarify on this further.

> I also wonder what happens when you have 2 levels of instead_of. If you have a @handle_urls decorator on B with instead_of=A, and on C with instead_of=B, is B used or is it C? Do we have a test for this?

That's an interesting point. I don't believe we have this in **web-poet** as the framework (e.g. **scrapy-poet**) is the one that should implement it. For **scrapy-poet**, I don't believe we test it explicitly on this. Created an issue in https://github.com/scrapinghub/scrapy-poet/issues/90 to keep track of it. Thanks!

Reply:
> I added some notes in the docstrings about this to remove the ambiguity and clarify on this further.

I do not see any change, did you push it yet?

Reply:
Oohh right sorry, forgot to push and mention it https://github.com/scrapinghub/web-poet/pull/84/commits/776cf0dd16a466504e64875307da86d4ee2a6115.